Fix OpenAI multimodal message logging support

mohammadsubhani opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/mohammadsubhani/mlflow/pull/16795?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16795/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16795/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16795/merge
```

</p>
</details>

  ### Related Issues/PRs

  <!-- Uncomment 'Resolve' if this PR can close the linked items. -->
  Resolve #16707

  ### What changes are proposed in this pull request?

  This PR fixes a TypeError that occurs when logging OpenAI models with multimodal
   message content (text + images). Previously, MLflow only supported string-based
   message content. When users attempted to log models with list-based multimodal
  content, the `_parse_format_fields` function would crash with "TypeError:
  expected str, got list".

  The fix implements recursive parsing logic to handle complex nested structures
  while maintaining full backward compatibility with existing string-based message
   templates.

  ### How is this PR tested?

  - [x] Existing unit/integration tests
  - [x] New unit/integration tests
  - [x] Manual tests

  **New Tests Added:**
  - `test_multimodal_messages()` - Tests multimodal content with template
  variables
  - `test_multimodal_messages_no_variables()` - Tests multimodal content without
  template variables
  - Both tests validate model saving, loading, and template variable extraction

  **Manual Testing:**
  - Validated fix works with real multimodal OpenAI message structures
  - Confirmed models can be saved and loaded successfully in MLflow UI
  - Tested with both Vision API format and custom multimodal formats

<img width="909" height="752" alt="Screenshot 2025-07-19 at 5 05 26 PM" src="https://github.com/user-attachments/assets/fdaa35e5-3782-4bb1-9a82-5da10fe6a0cb" />

<img width="2337" height="485" alt="Screenshot 2025-07-19 at 5 05 50 PM" src="https://github.com/user-attachments/assets/a79ef83c-83b2-443f-a5ad-f98c7aa6d35b" />


  ### Does this PR require documentation update?

  - [x] No. You can skip the rest of this section.

  ### Release Notes

  #### Is this a user-facing change?

  - [x] Yes. Give a description of this change to be included in the release notes
   for MLflow users.

  **Description:** MLflow now supports logging OpenAI models with multimodal
  messages containing both text and image content. Previously, attempting to log
  such models would result in a TypeError. This enhancement enables users to work
  with OpenAI's Vision API and other multimodal chat completion models while
  maintaining full backward compatibility.

  #### What component(s), interfaces, languages, and integrations does this PR
  affect?

  Components:
- [x] `area/models`: MLmodel format, model
  serialization/deserialization, flavors

  Interface:

  - [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript,
  JavaScript dev server
  - [ ] `area/docker`: Docker use across MLflow's components, such as
  MLflow Projects and MLflow Models
  - [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or
   Model Registry
  - [ ] `area/windows`: Windows support

  Language:

  - [ ] `language/r`: R APIs and clients
  - [ ] `language/java`: Java APIs and clients
  - [ ] `language/new`: Proposals for new client languages

  Integrations:

  - [ ] `integrations/azure`: Azure and Azure ML integrations
  - [ ] `integrations/sagemaker`: SageMaker integrations
  - [ ] `integrations/databricks`: Databricks integrations

  <a name="release-note-category"></a>

  #### How should the PR be classified in the release notes? Choose one:

  - [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

  #### Should this PR be included in the next patch release?

  - [x] Yes (this PR will be cherry-picked and included in the next patch release)

Subhanimuhammed committed:
Fix OpenAI multimodal message logging support

Fixed TypeError when logging OpenAI models with multimodal messages (text + images).
The _parse_format_fields and format_chat methods now handle list/dict content structures
recursively, maintaining backward compatibility with string-based messages.

- Enhanced _parse_format_fields to recursively parse multimodal content structures
- Updated format_chat to format nested list/dict content using recursive approach
- Added comprehensive tests for multimodal messages with and without variables
- Maintains full backward compatibility with existing string-based message templates

Fixes #16707

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>

Subhanimuhammed committed:
Improve documentation with Google-style docstrings for multimodal functions

Added comprehensive Google-style docstrings to enhance code documentation:

- _parse_format_fields(): Added detailed docstring explaining recursive parsing
  of template variables from various content types with examples
- format_chat(): Added comprehensive docstring explaining multimodal message
  formatting with parameter substitution examples
- format_value(): Added clear docstring for recursive template substitution
- test_multimodal_messages(): Added detailed test documentation explaining
  multimodal content testing with template variables
- test_multimodal_messages_no_variables(): Added comprehensive test documentation
  for static multimodal content handling

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>

Subhanimuhammed committed:
Add explicit type hints to recursive functions for better static analysis

Enhanced type safety and readability with parameterized generic types:

- _parse_format_fields(): Added Union[str, list[Any], dict[str, Any], Any] type hints
- format_value(): Added explicit parameter and return type annotations

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>

B-Step62 commented:
@mohammadsubhani The fix looks reasonable, but can you remove verbose docstrings in the private method and test cases?

Subhanimuhammed committed:
Remove verbose docstrings from private methods and test cases

- Simplify _parse_format_fields docstring to concise one-liner
- Remove verbose docstring from format_chat method
- Remove docstring from private format_value helper function
- Remove verbose docstrings from multimodal test functions

Addresses reviewer feedback to reduce verbosity in private methods
and test cases while maintaining clear, functional code.

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>

mohammadsubhani commented:
I have implemented the changes @B-Step62 , i.e. removed verbose doc strings from private methods as well as in the test cases. Can you kindly re-review it? Thank you!

B-Step62 reviewed (approved):
LGTM! Thank you for proactively filing the fix!
